### PR TITLE
graph-builder: add support for configuration file 

### PR DESCRIPTION
**Pull request by lucab**
This reworks plugins setup and graph-builder CLI parser to allow for sourcing complex/structured options from configuration file(s). In particular, configuration logic now cares about the following entities:
 * "options": configuration fragments (CLI flags, file snippets), optional and stringly-typed.
 * "unified config": configuration document, result of merging all options and defaults.
 * "app settings": runtime settings, result of config validation.

Ref: https://jira.coreos.com/browse/CORS-967
/cc @steveeJ

### CODE REVIEW

**Comment by lucab** [comment]
Rebased to master and added graph-builder configuration options for https://github.com/openshift/cincinnati/pull/72#issuecomment-473327021.

File: graph-builder/src/main.rs
Comment (lucab) [inline graph-builder/src/main.rs]: Self note: s/opts/settings/.

File: cincinnati/src/plugins/registry.rs
Comment (lucab) [inline cincinnati/src/plugins/registry.rs]: Self note: check if there is a better term instead of "registry" (which could be confusing due to docker-registry).

**Comment by lucab** [comment]
@steveeJ I've dropped all the plugins logic from here as requested, and split that to https://github.com/openshift/cincinnati/pull/74.

File: graph-builder/src/config/unified.rs
Comment (steveej) [inline graph-builder/src/config/unified.rs]: Is there a reason for why we can't have `Option`s in here?

File: graph-builder/src/config/unified.rs
Comment (steveej) [inline graph-builder/src/config/unified.rs]: how about `assign_if_some` here? the second `_` is emitable if you think it's too long. I think it would be a more intuitive name, as, at least for me, _maybe_ didn't tell me what the condition is.

File: graph-builder/src/config/unified.rs
Comment (steveej) [inline graph-builder/src/config/unified.rs]: s/Eaarly/Early/

File: graph-builder/src/config/unified.rs
Comment (steveej) [inline graph-builder/src/config/unified.rs]: Are the subsequent plugin options leftover in this PR or required?

File: graph-builder/src/graph.rs
Comment (steveej) [inline graph-builder/src/graph.rs]: The condition for gating the plugins got lost here

File: graph-builder/src/config/cli.rs
Comment (lucab) [reply graph-builder/src/config/cli.rs]: I don't plan to reuse this; however, I'm bad at naming things :(

This is technically a "DockerRegistryV2" upstream, and there are likely a "[OciDistributionVx](https://github.com/opencontainers/distribution-spec)" and a "FedoraCoreos" coming up at some point in the future. I went for `registry` only because it was shorter for options.

File: graph-builder/src/config/file.rs
Comment (lucab) [reply graph-builder/src/config/file.rs:42]: Not directly because the expected error type is a `failure::Error` while this is a `failure::Context`. I could chain some error-mapping to massage types, but I think it's more straightforward to read in its current form.

File: graph-builder/src/config/file.rs
Comment (lucab) [reply graph-builder/src/config/file.rs:61]: For some sections I think we can, yes. Top-level and plugins will be separate, but the others can be shared and derive `Debug + Deserialize + StructOpt` in an common `options.rs`.

File: graph-builder/src/config/unified.rs
Comment (lucab) [reply graph-builder/src/config/unified.rs]: Is there anything specific that you think should be an `Option`?

This struct holds configuration entries, after merging together defaults+toml+cli. Unless something here requires a real `None`, most of the entries will have real values (either from user-options or from defaults).

File: graph-builder/src/graph.rs
Comment (lucab) [reply graph-builder/src/graph.rs]: This is currently unused in our deployment, and #74 takes care of that (plugins not in the config are not enabled). I dropped the CLI flag here.

**Comment by lucab** [comment]
@steveeJ thanks for the review, I've taken most of your suggestions and amended the code. In particular, CLI and TOML are now sharing their types, and I've split a minor enhancement to https://github.com/openshift/cincinnati/pull/76. PTAL.

File: graph-builder/src/config/unified.rs
Comment (steveej) [reply graph-builder/src/config/unified.rs]: I would imagine that this struct could be a mere composition of the ones we already have for CLI and file, so this would resemble the same structure everywhere. I understand this places the burden of handling the `Option`s again in the code :thinking:

File: graph-builder/src/config/file.rs
Comment (steveej) [reply graph-builder/src/config/file.rs:61]: Now `UpstreamOptions` is only available for file configuration, correct?

File: graph-builder/src/config/file.rs
Comment (steveej) [reply graph-builder/src/config/file.rs:42]: So the `?` handles that conversion automatically?

**Comment by lucab** [comment]
@steveeJ I did a major rework on the PR in the direction of https://github.com/openshift/cincinnati/pull/69#discussion_r268681483. This means that the `Unified` struct does not exist anymore, input parsing goes via structopt+serde, and most of the structs are re-used.

While this definitely still needs some polishing, can you please another pass to see if suits your taste better than the previous iteration?

File: graph-builder/src/config/options.rs
Comment (lucab) [inline graph-builder/src/config/options.rs:72]: Self-note: this only works on toolchain >= 1.32.

File: graph-builder/src/config/settings.rs
Comment (steveej) [inline graph-builder/src/config/settings.rs:21]: Might it simplify/streamline the rest of the code if this is wrapped in an `Option` too?

File: graph-builder/src/config/settings.rs
Comment (steveej) [inline graph-builder/src/config/settings.rs]: The comment below will probably influence this, but in the current stat we could use `Default::default()` here

File: graph-builder/src/config/settings.rs
Comment (steveej) [inline graph-builder/src/config/settings.rs]: `Default::default()`?

File: graph-builder/src/config/file.rs
Comment (steveej) [reply graph-builder/src/config/file.rs:61]: For reference: we've agreed that we don't want to complicate the CLI flags more and thus keep this only in the config file for now. We might revisit later.

File: graph-builder/src/graph.rs
Comment (steveej) [reply graph-builder/src/graph.rs]: I needed the flag for testing the integration with Satellite. That might come up again anytime soon so I'm in favor to leave something in to gate this.

**Review by steveej — COMMENTED** [review]
The overall approach looks very clean to me, especially the simplicity of the `assemble` function.
I would've wished for some code generation (macros) for the merge implementations but I wouldn't block on that if there exists no ready-made solution right now.

I have added a couple of comments to be addressed and there are also some leftover from the previous review. Thanks for exploring this so thoroughly!

File: graph-builder/src/config/options.rs
Comment (lucab) [reply graph-builder/src/config/options.rs:72]: Split to https://github.com/openshift/cincinnati/pull/77.

File: graph-builder/src/config/settings.rs
Comment (lucab) [reply graph-builder/src/config/settings.rs]: Here and above, `Default::default()` is implicit if I remove the annotation (but I liked it explicit). Do you prefer me to drop all the non-custom default annotations?

File: graph-builder/src/config/options.rs
Comment (steveeJ) [inline graph-builder/src/config/options.rs]: need a docstring here

File: graph-builder/src/config/mod.rs
Comment (steveej) [inline graph-builder/src/config/mod.rs]: need a docstring for this

File: graph-builder/src/config/settings.rs
Comment (lucab) [reply graph-builder/src/config/settings.rs]: Discussed out-of-band: let's use the implicit form.

File: graph-builder/src/config/options.rs
Comment (steveej) [reply graph-builder/src/config/options.rs:72]: So a rebase should fix the build failure :+1:

File: graph-builder/src/config/settings.rs
Comment (lucab) [reply graph-builder/src/config/settings.rs:21]: Not really, it will even make set-appending nested into an additional level of option-matching.

**Comment by lucab** [comment]
I added the final tests and documentation, completing the last TODOs. Dropped the WIP label, this is ready for the last round of review.

File: docs/user/graph-builder-configuration.md
Comment (steveej) [inline docs/user/graph-builder-configuration.md]: should we rename this now while we're at it? to either "pause" or "delay"?

**Review by steveej — DISMISSED** [review]
thanks for the nice rework!

there's a small doc change I'd like to make and a couple of questions/suggestions I added (please feel free to dismiss the suggestions).

File: docs/user/graph-builder-configuration.md
Comment (lucab) [reply docs/user/graph-builder-configuration.md]: If you are fine with mixing the rename in here, I'm happy to do that. I also wanted to mention the time unit here. What about `pause_secs`?

File: docs/user/graph-builder-configuration.md
Comment (lucab) [reply docs/user/graph-builder-configuration.md:1]: I think it is technically feasible somehow (worst case, via sed), but I am not aware of a clean way to do it right now. I'll chalk this to a ticket for later.

File: docs/user/graph-builder-configuration.md
Comment (steveej) [reply docs/user/graph-builder-configuration.md]: `pause_secs` seems fine

File: graph-builder/src/config/file.rs
Comment (lucab) [reply graph-builder/src/config/file.rs:42]: Eventually everything converts to a `Result` and `std::error::Error` trait. There a few layers to go through in `failure`, though. I don't know the exact path that `?` goes through, but it is pretty effective at doing those conversions automatically, yes.

File: docs/user/graph-builder-configuration.md
Comment (lucab) [reply docs/user/graph-builder-configuration.md:1]: Moved to https://jira.coreos.com/browse/CIN-21

**Comment by lucab** [comment]
Pushed another update addressing last comments, PTAL

**Review by steveej — APPROVED** [review]
:tada: 

/lgtm

**Pull request merged**
